packages/contracts/src/Inventory.sol

```diff
@@ -245,22 +294,44 @@ contract Inventory is Ownable {
     // ---------------------------------------------------------------------------------------------
 
     // Checks that the player has all the cards in the given deck in the inventory.
-    function checkDeck(address player, uint8 deckID) external view exists(player, deckID) {
-        Deck storage deck = decks[player][deckID];
-        for (uint256 i = 0; i < deck.cards.length ; ++i) {
-            uint256 cardID = deck.cards[i];
-            if (inventoryCardsCollection.ownerOf(cardID) != player)
-                revert CardNotInInventory(cardID);
+    function checkDeck(address player, uint8 deckID) 
+        external 
+        view 
+        exists(player, deckID) 
+    {
+        Deck memory deck = decks[player][deckID];
+        uint256[] memory sortedCards = Utils.sort(deck.cards);
+        uint256 deckLength = sortedCards.length;
+        uint256 cardCopies;
+        uint256 prev;
+        for (uint256 i = 0; i < deckLength; ++i) {
+            uint256 cardID = sortedCards[i];
+            if(cardID == prev){ 
+                cardCopies++;
+                // check that each card does not exceed its maximum amount of copies.
+                if (cardCopies > MAX_CARD_COPY){
+                    revert CardExceedsMaxCopy(cardID);
+                }
+            } else {
+                if (inventoryCardsCollection.ownerOf(cardID) != player){
+                    revert CardNotInInventory(cardID);
+                }
+                cardCopies = uint256(1); 
+            }
+            prev = sortedCards[i];
```

Comment:
This can be in the `else` branch.

Reply:
hmm, i dont understand.
if i put it in the `else` block then it would only be executed conditionally whereas it needs to be executed on each iteration.
what am i missing? 👀

Reply:
It does not need to be executed in each iteration, only when the card changes. If you're in the `if` branch, then you know the card is the same as last one, so `prev` can remain the same. In the `else` branch is when the card changes and `prev`needs to be updated.

Reply:
true true
got it!